File: pkg/worker/request.go
package worker

import (
	"encoding/binary"

	"github.com/HughNian/nmid/pkg/model"
	"github.com/HughNian/nmid/pkg/utils"
)

type Request struct {
	DataType uint32
	Data     []byte
	DataLen  uint32

	Handle     string
	HandleLen  uint32
	ParamsType uint32
	ParamsLen  uint32
	Params     []byte
	JobId      string
	JobIdLen   uint32
	Ret        []byte
	RetLen     uint32
}

func NewReq() (req *Request) {
	req = &Request{
		Data:      make([]byte, 0),
		DataLen:   0,
		Handle:    ``,
		HandleLen: 0,
		ParamsLen: 0,
		Params:    make([]byte, 0),
		Ret:       make([]byte, 0),
		RetLen:    0,
	}
	return
}

// HeartBeatPack 打包内容-心跳
func (req *Request) HeartBeatPack() (content []byte, err error) {
	data := "PING"

	req.DataType = model.PDT_W_HEARTBEAT_PING
	req.DataLen = uint32(len(data))
	req.Data = []byte(data)
	content = req.Data

	return
}

// SetWorkerName 打包内容-设置worker名称
func (req *Request) SetWorkerName(workerName string) (content []byte, err error) {
	req.DataType = model.PDT_W_SET_NAME
	req.DataLen = uint32(len(workerName))
	req.Data = []byte(workerName)
	content = req.Data

	return
}

// AddFunctionPack 打包内容-添加方法
func (req *Request) AddFunctionPack(funcName string) (content []byte, err error) {
	req.DataType = model.PDT_W_ADD_FUNC
	req.DataLen = uint32(len(funcName))
	req.Data = []byte(funcName)
	content = req.Data

	return
}

// DelFunctionPack 打包内容-删除方法
func (req *Request) DelFunctionPack(funcName string) (content []byte, err error) {
	req.DataType = model.PDT_W_DEL_FUNC
	req.DataLen = uint32(len(funcName))
	req.Data = []byte(funcName)
	content = req.Data

	return
}

// GrabDataPack 打包内容-抓取任务
func (req *Request) GrabDataPack() (content []byte, err error) {
	req.DataType = model.PDT_W_GRAB_JOB
	req.DataLen = 0
	req.Data = []byte(``)
	content = req.Data

	return
}

// WakeupPack 打包内容-唤醒
func (req *Request) WakeupPack() {
	req.DataType = model.PDT_WAKEUP
	req.DataLen = 0
	req.Data = []byte(``)
}

// LimitExceedPack 打包内容-限流
func (req *Request) LimitExceedPack() {
	req.DataType = model.PDT_RATELIMIT
	req.DataLen = 0
	req.Data = []byte(``)
}

// RetPack 打包内容-返回结果
func (req *Request) RetPack(ret []byte) (content []byte, err error) {
	req.Ret = ret
	req.RetLen = uint32(len(ret))

	req.DataType = model.PDT_W_RETURN_DATA
	req.DataLen = model.UINT32_SIZE + req.HandleLen + model.UINT32_SIZE + req.ParamsLen + model.UINT32_SIZE + req.RetLen + model.UINT32_SIZE + req.JobIdLen

	length := int(req.DataLen)
	content = utils.GetBuffer(length)
	binary.BigEndian.PutUint32(content[:model.UINT32_SIZE], req.HandleLen)
	start := model.UINT32_SIZE
	end := int(model.UINT32_SIZE + req.HandleLen)
	copy(content[start:end], []byte(req.Handle))
	start = end
	end = start + model.UINT32_SIZE
	binary.BigEndian.PutUint32(content[start:end], uint32(req.ParamsLen))
	start = end
	end = start + int(req.ParamsLen)
	copy(content[start:end], req.Params)
	start = end
	end = start + model.UINT32_SIZE
	binary.BigEndian.PutUint32(content[start:end], req.RetLen)
	start = end
	end = start + int(req.RetLen)
	copy(content[start:end], req.Ret)
	start = end
	end = start + model.UINT32_SIZE
	binary.BigEndian.PutUint32(content[start:end], req.JobIdLen)
	start = end
	end = start + int(req.JobIdLen)
	copy(content[start:end], req.JobId)
	req.Data = content

	return
}

// EncodePack 打包
func (req *Request) EncodePack() (data []byte) {
	len := model.MIN_DATA_SIZE + req.DataLen //add 12 bytes head
	data = utils.GetBuffer(int(len))

	binary.BigEndian.PutUint32(data[:4], model.CONN_TYPE_WORKER)
	binary.BigEndian.PutUint32(data[4:8], req.DataType)
	binary.BigEndian.PutUint32(data[8:model.MIN_DATA_SIZE], req.DataLen)
	copy(data[model.MIN_DATA_SIZE:], req.Data)

	return
}

File: sdk/python3/nmidsdk/worker/request.py
import struct
import nmidsdk.model.const as const

class Request:
    def __init__(self):
        self.data_type = 0
        self.data = bytearray()
        self.data_len = 0
        self.handle = ""
        self.handle_len = 0
        self.params_type = 0
        self.params_handle_type = 0
        self.params_len = 0
        self.params = bytearray()
        self.job_id = ""
        self.job_id_len = 0
        self.ret = bytearray()
        self.ret_len = 0

    def heart_beat_pack(self):
        data = "PING"
        self.data_type = const.PDT_W_HEARTBEAT_PING
        self.data_len = len(data)
        self.data = data.encode()
        
        return self.data

    def set_worker_name(self, worker_name: bytearray) -> bytearray:
        self.data_type = const.PDT_W_SET_NAME
        self.data_len = len(worker_name)
        self.data = worker_name

        return self.data
    
    def add_function_pack(self, func_name: bytearray) -> bytearray:
        self.data_type = const.PDT_W_ADD_FUNC
        self.data_len = len(func_name)
        self.data = func_name

        return self.data
    
    def del_function_pack(self, func_name: bytearray) -> bytearray:
        self.data_type = const.PDT_W_DEL_FUNC
        self.data_len = len(func_name)
        self.data = func_name

        return self.data
    
    def grab_data_pack(self) -> bytearray:
        self.data_type = const.PDT_W_GRAB_JOB
        self.data_len = 0
        self.data = bytearray()

        return self.data
    
    def wakeup_pack(self) -> bytearray:
        self.data_type = const.PDT_WAKEUP
        self.data_len = 0
        self.data = bytearray()

        return self.data
    
    def limit_exceed_pack(self) -> bytearray:
        self.data_type = const.PDT_RATELIMIT
        self.data_len = 0
        self.data = bytearray()

        return self.data
    
    def ret_pack(self, ret: bytes) -> bytearray:
        self.ret = ret
        self.ret_len = len(ret)

        self.data_type = const.PDT_W_RETURN_DATA
        self.data_len = const.UINT32_SIZE + self.handle_len + const.UINT32_SIZE + self.params_len + const.UINT32_SIZE + self.ret_len + const.UINT32_SIZE + self.job_id_len

        length = int(self.data_len)
        content = bytearray(length)

        # 序列化数据
        content[:const.UINT32_SIZE] = struct.pack('>I', self.handle_len)
        content[const.UINT32_SIZE:const.UINT32_SIZE + self.handle_len] = self.handle.encode()
        content[const.UINT32_SIZE + self.handle_len:const.UINT32_SIZE + self.handle_len + const.UINT32_SIZE] = struct.pack('>I', self.params_len)
        content[const.UINT32_SIZE + self.handle_len + const.UINT32_SIZE:const.UINT32_SIZE + self.handle_len + const.UINT32_SIZE + self.params_len] = self.params
        content[const.UINT32_SIZE + self.handle_len + const.UINT32_SIZE + self.params_len:const.UINT32_SIZE + self.handle_len + 2*const.UINT32_SIZE + self.params_len] = struct.pack('>I', self.ret_len)
        content[const.UINT32_SIZE + self.handle_len + 2*const.UINT32_SIZE + self.params_len:const.UINT32_SIZE + self.handle_len + 2*const.UINT32_SIZE + self.params_len + self.ret_len] = self.ret
        content[const.UINT32_SIZE + self.handle_len + 2*const.UINT32_SIZE + self.params_len + self.ret_len:const.UINT32_SIZE + self.handle_len + 3*const.UINT32_SIZE + self.params_len + self.ret_len] = struct.pack('>I', self.job_id_len)
        content[const.UINT32_SIZE + self.handle_len + 3*const.UINT32_SIZE + self.params_len + self.ret_len:const.UINT32_SIZE + self.handle_len + 3*const.UINT32_SIZE + self.params_len + self.ret_len + self.job_id_len] = self.job_id.encode()

        self.data = content
        return content

    def encode_pack(self) -> bytearray:
        len = const.MIN_DATA_SIZE + self.data_len  # add 12 bytes head
        data = bytearray(len)

        # 添加头部信息
        data[:4] = struct.pack('>I', const.CONN_TYPE_WORKER)
        data[4:8] = struct.pack('>I', self.data_type)
        data[8:const.MIN_DATA_SIZE] = struct.pack('>I', self.data_len)
        data[const.MIN_DATA_SIZE:] = self.data

        return data